### PR TITLE
apf: introduce final seats

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

This PR implements the `final seats` concept, see implements https://github.com/kubernetes/kubernetes/pull/103539#issuecomment-925995530

#### Which issue(s) this PR fixes:
Fixes #

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by tkashem** [comment]
work on tests is in progress

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go:112]: Units confusion here.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go:114]: I think only two things need to be tracked:
- the sum of the max-width, for the purpose of computing dR/dt
- the sum of the work, for the purpose of scoring in the shuffle-sharding

BTW, "work estimate" is not the best name for the pair of those two quantities.  Maybe they do not need to be put together in one struct.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/width.go
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/width.go:37]:
```suggestion
	// InitialSeats is the number occupied while the server is executing the request
```

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/width.go
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/width.go:40]:
```suggestion
	// FinalSeats is the number occupied at the end, during the AdditionalLatency
```

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/width.go
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/width.go:50]: Let's call this `MaxSeats`

**Review by MikeSpreitzer — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/width.go
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/width.go:36]: I am becoming less satisfied with the name "WorkEstimate" for this datatype.  For one thing, the word "estimate" is not very helpful here.  It is the _usage_ of this data that determines whether it is estimated or not.  More importantly, this struct does not simply hold "work" but rather components that go into computing work.  The work of a request is now a function of four things (two widths and two durations), three of which are here and one of which is in the queueset (thus far).

**Comment by MikeSpreitzer** [comment]
@MikeSpreitzer

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/fifo_list_test.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/fifo_list_test.go:178]: I agree, microseconds-seats with int64 will simplify the code and tests. I am going to make the changes here and then will rebase on top of #105322

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/fifo_list_test.go
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/fifo_list_test.go:157]: nit: how about:
```
multiplier := (-)1 (based on add - or even pass explicitly to the function)
we.InitialSeatsSum += multiplier * req.InitialSeats()
...
```

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/fifo_list_test.go
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/fifo_list_test.go:170]: if !reflect.DeepEqual(want, got) {
  t.Errorf("Unexpected work estimate: %v", expected: %v", got, want)
}

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:333]: This sum has mixed units.  The left term is seat-seconds, the right term is seat-microseconds.

Why introduce microseconds when the code was previously consistently using seconds?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go:107]: "Requested" is not doing much useful work here.  We do _not_ have a distinction between, say, "requested" vs "granted" seats.

A more useful thing to mention in this field's name and comment is that we are talking about summing the _max_ seats.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go:109]: "duration" here is wrong, it is "work".

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/width.go
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/width.go:41]: I still think it would be good to update this comment to be explicit about what "final" actually means.  This is the  place where the initial/final distinction originates, so this is the best place to document what "initial" and "final" correspond to.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:515]: I think the score for a queue should be the sum, over all the executing and waiting requests, of the estimated work in that request (`req.WorkEstimate.InitialSeats * qs.estimatedServiceSeconds + req.WorkEstimate.FinalSeats * req.WorkEstimate.AdditionalLatency.Seconds()`).

The current draft of this PR make a different calculation.  It uses the sum over executing and waiting requests of `req.MaxSeats() * qs.estimatedServiceSeconds` plus the sum over waiting requests of `req.WorkEstimate.FinalSeats * req.WorkEstimate.AdditionalLatency.Seconds()`.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:860]: This PR changes _some_ of the updates to `queue::virtualStart` to use units of seat-microseconds, leaving others (like this one) in units of seat-seconds.  I see no need for this PR to introduce any switch to microseconds, since we are already contemplating an additional PR to change how work is represented.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:333]: yes, waiting to rebase on top of https://github.com/kubernetes/kubernetes/pull/105322.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:860]: yes, in anticipation of https://github.com/kubernetes/kubernetes/pull/105322, once #105322 gets merged I will rebase.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:333]: Please now rebase on top of https://github.com/kubernetes/kubernetes/pull/105412

All these functions should now simply return SeatSeconds

File: staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go]: done

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go]: renamed to `MaxSeats`

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:824]: if we define `FinalSeats` as the number of seats occupied at the end, during the `AdditionalLatency` then:
- `FinalSeats <= InitialSeats`:  we release `InitialSeats-FinalSeats` immediately after the request finishes and then hold `FinalSeats` for `AdditionalLatency`

- `FinalSeats > InitialSeats`: after the request finishes, we will need to acquire `FinalSeats - InitialSeats` more seats from the pool which is not a guaranteed process.

This PR now reserves `MaxSeats` when the request is dispatched, but this seems like an under utilization when `FinalSeats > InitialSeats`

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:824]: - Since `FinalSeats < InitialSeats` never actually happens, I would not worry about supporting it.
- Remember that when `InitialSeats < FinalSeats`, while this guarantees seats will not be occupied it does not guarantee under-utilization because our seats are just abstractions that are multiplexed onto the underlying real resources.  For example, when multiplexing fewer processes onto a fixed number of CPUs, those processes get work done faster (assuming CPU is the bottleneck and there are at least as many processes as CPUs).
- We have discussed the potential for under-utilization in review comments and meetings, and the current thinking is to let the first draft take no particular steps to combat it and see how that works out.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:523]: > I think that the score used for shuffle-sharding should be the sum of the work in the queue's executing and waiting requests.

I'm not sure I agree. For requests that are already executing, they may potentially be almost finished already, in which case your algorithm will look this queue much worse than it can be in practice.
We should try to project the expected remaining amount of work instead (which we can do based on the R() meter and the value.


That said - if we assume we want to have what you want (which I'm temporarily fine as this is the logic we have currently anyway) - then it should be:

SeatsTimesDuration(queueSum.InitialSeatsSum(), qs.estimatedServiceDuration) + queueSum.AdditionaSeatsSecondsSum + queue.seatsInUse * something (which we actually can't say what it is, because we're not tracing in which phase a request is - so assuming qs.estimatedServiceDuration is somewhat reasonable guess).

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go]: `SeatsTimesDuration` is covered by its own set of tests, besides if i wanted to use a constant the test would have to know about `ssScale`, no?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:523]: okay, fixed.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:523]: I would personally suggest going with the first option from the above list and merging this PR.

I agree it's not ideal - but given it's roughly what we had so far, I don't perceive it as a regression, even though it's not exactly what we want.

This would unblock this PR and let us make further progress, and we can discuss in the meantime how we can improve that further.

@MikeSpreitzer - WDYT?

**Review by wojtek-t — COMMENTED** [review]
The opened comment is my one remaining one at this point.

**Comment by tkashem** [comment]
> This LGTM, but I would like to understand test failure - this isn't failing for the reason you pasted.

i debugged further by commenting some `t.Log` lines and was able to trace the failure to 
```
$ stress ./queueset.test -test.v -test.run=TestDifferentWidths

=== RUN   TestDifferentWidths
    queueset_test.go:211: 2021-10-07 10:50:50.454843113: Start TestDifferentWidths, doSplit=false, clk=0xc0002a2200, grc=0xc0002be404
    queueset_test.go:308: 2021-10-07 10:51:10.454843113: End
    queueset_test.go:336: TestDifferentWidths client 0 last=true expectFair=true margin=0.125 got an Average of 2.5999999960000006 but the expected average was 3
    queueset_test.go:338: TestDifferentWidths client 1 last=true expectFair=true margin=0.125 got an Average of 3.099999998 and the expected average was 3

```
it's originating from here - https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go#L336. it fails in expected fair allocation not being met.

interestingly, if i run `TestDifferentWidths` serially 100 times i can't repro this error. I think with `stress` running the test binary in parallel (8 by default) it stretches CPU and adds latency in the request execution duration. that causes the test to fail.

also, `stress` captures the log from these test binaries running in parallel using a channel and it somehow captures only a portion of the test log, that's why after removing the `t.Log` statements from the test i was able to see the actual test error log.

**Pull request merged**
